docs/design/batch-request.md

```diff
@@ -0,0 +1,260 @@
+|  -32007   | Method `<method>` is not permitted as part of batch requests  | Execute method on single request |
+
+
+
+### Errors on a Batch Item
+
+This happens when there are errors on one or many individual requests items of the request array.
+
+1. The request item method does not exist (Method not found)
+2. The request item does not conform to the JSON-RPC 2.0 specification (Invalid Request)
+3. Missing required parameters or wrong parameter type 
+4. The request item is valid but the result is an error, like an `eth_call` that returns a `revert` or `out of gas` error
+
+Example of a response with the above errors (in the same order):
+
+```json
+[
+    {
+        "error": {
+            "message": "Method non_existent_method not found",
+            "code": -32601
+        },
+        "jsonrpc": "2.0",
+        "id": 1
+    },
+    {
+        "error": {
+            "message": "Invalid Request",
+            "code": -32600
+        },
+        "jsonrpc": "2.0",
+        "id": null
+    },
+    {
+        "error": {
+            "code": -32602,
+            "name": "Missing required parameters",
+            "message": "[Request ID: 9e2c40fb-3406-4c48-b7ac-7a40e9bdb880] Missing value for required parameter 1"
+        },
+        "jsonrpc": "2.0",
+        "id": 3
+    },
+    {
+        "error": {
+            "code": -32005,
+            "name": "gasLimit too high",
+            "message": "[Request ID: 9e2c40fb-3406-4c48-b7ac-7a40e9bdb880] Transaction gas limit '0xe1f21c67' exceeds block gas limit '15000000'"
+        },
+        "jsonrpc": "2.0",
+        "id": 4
+    }
+]
+```
+
+### Bad Request on the Batch Request as a whole
+This type of error happens when the whole batch request fails and could happen for the following reasons:
+1. The request is not a valid JSON object or Array of objects
+```json
+{
+    "error": {
+        "message": "Parse error",
+        "code": -32700
+    },
+    "jsonrpc": "2.0",
+    "id": null
+}
+```
+2. The feature is disabled
+```json
+{
+    "error": {
+        "code": -32202,
+        "name": "Batch requests disabled",
+        "message": "Batch requests are disabled"
+    },
+    "jsonrpc": "2.0",
+    "id": null
+}
+```
+3. The request Array size is bigger than the configured limit
+
+```json
+{
+    "jsonrpc": "2.0",
+    "error": {
+        "code": -32005,
+        "message": "batch item count exceeded"
+    }
+}
+```
+
+## Metrics Capture Enhancement
+Currently metrics are captured using a KoaApp middleware that captures the metric `rpc_relay_method_response` per each HTTP request capturing the amount of time it took to process the request in milliseconds (ms).
+
+The metric currently uses the following labels:
+
+1. `method` (method name such as eth_call, eth_getBlockByNumber or batch_request)
+2. `status` (HTTP status code)
+
+This approach works for both single and batch requests, however it does not capture the metrics for each individual request within the batch request.
```

Comment:
Q: Why does this not work. Could the metric not be captured in the batching logic so that there's no need for a new metric?
Is there no share request processing logic that would allow the metric to be captured regardless of if it was part of a single or batch request?

Reply:
yes, is possible to use the current metric, and only use `1` but then it won't be the same as what we are measuring right now (we are mostly using that metric to measure `http latency` to find potential `timeouts`. 

But come to think of it again, I think we could use it and just add a new label, isBatched, and we can filter out all the occurrences that come with this flag when doing a dashboard for `http latency` and not when measuring the time it takes a method to return...

I like this approach best 👍

Reply:
✅ I was able to find another approach that might also work better and simpler 👍